cli:connectivity: fix egress dev on aws-cni for pod-to-pod-encryption-v2

smagnani96 opened the pull request:
Partially fixes (only for v2): https://github.com/cilium/cilium/issues/37675

Fixes: https://github.com/cilium/cilium/issues/37682

```release-note
Fix egress device computation in cli connectivity pod-to-pod-encryption-v2 tests for AWS chaining mode.
```

smagnani96 committed:
cli:connectivity: fix egress dev on aws-cni for pod-to-pod-encryption-v2

This commit adjusts the current logic in `pod-to-pod-encryption-v2` for
retrieving the egress device from client-server pods and vice-versa.
Before the patch, we solely rely on `ip route get dstIP`, using the
returned device as egress. However, we noticed that in aws-cni chaining
mode, such returned device is not actually used. Instead, looking at the
routing tables in the server node, we found that there's a specific table
with a route that uses a different egress device for that particular IP.

Therefore, the final command to be issued should be:
`ip route get {dstIP} from {srcIP} iif cilium_host`.
The interface specified does not matter since, if present, the route
matching srcIP is not making use of the interface field to match.

The related issue in our previous tests is #37675.

We now fix this for pod-to-pod-encryption-v2, but we might want to adopt
the same logic for v1.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

ldelossa commented on cilium-cli/connectivity/tests/encryption_v2.go:
```diff
@@ -110,52 +110,138 @@ func (s *podToPodEncryptionV2) Name() string {
+	//
+	// example json output:
+	// [{"dst":"default","gateway":"192.168.128.1","dev":"ens6","flags":[]},{"dst":"192.168.128.1","dev":"ens6","scope":"link","flags":[]}]
+	tableIPRouteGet := func(table string) (string, error) {
```
This function is specifically looking for the 'device for the default route' but comments and function name don't really mention this.

smagnani96 replied:
You're absolutely right, didn't update comments in the code after changing the logic :sweat_smile: 
Just adjusted the comments, thanks!

ldelossa reviewed (approved):
Other then my comment, this looks good.

pchaigno commented on cilium-cli/connectivity/tests/encryption_v2.go:
```diff
@@ -114,52 +114,131 @@ func (s *podToPodEncryptionV2) Name() string {
 // network namespace used to send traffic to dst.
-func (s *podToPodEncryptionV2) resolveEgressDevice(ctx context.Context, srcHostNS *check.Pod, dst *check.Pod) (string, error) {
+func (s *podToPodEncryptionV2) resolveEgressDevice(ctx context.Context, srcHostNS *check.Pod, src, dst *check.Pod) (string, error) {
+	// deviceFromIPRoute issues `ip route get` for destination in provided host network
```
Did you try using `ip route get ... from podIP`? AFAICT, that would remove the need for the two functions and for iterating over ip rules... or maybe I'm missing something?

ldelossa replied:
@pchaigno so the `from` argument is news to me, TIL. 

I'm assuming you are saying do something like `ip rout get {dst} from {src_ip}` and `ip route get` will do the necessary ip rule and subsequent table lookup?

pchaigno replied:
Yep, exactly.

smagnani96 replied:
Thanks @pchaigno for the point. With Louis, we verified your saying in an EKS cluster, and the command seems to need also an interface specified (can be whatever).
Going to change our logic to 1 single lookup `ip route get {dst} from {src_up} iif cilium_host` (cilium_host will always be present).

smagnani96 replied:
Just updated the commits.
Re-introduced commit 3 to re-run `/ci-awscni` with debug/verbose output + forcing `client3` during CLI tests. Won't be merged.

ldelossa replied:
Just to add to that... 

`ip route get from ...` requires an `iif` (input interface) to return anything useful. 
ENI ip rules do not specify an `iif` specifically in the ip rule, therefore ANY interface works, as packets sourced from `src_ip` from ANY interface will result in the `ip rule` being triggered. 
Therefore, just provide `cilium_host` per what Simone mentions, its always there. 

Nice catch and suggestion Paul, that makes things way simpler :)

pchaigno replied:
Uh, I didn't need to specify an interface when I checked locally :thinking: Anyway, if any interface works the same, I guess it doesn't matter.

ldelossa replied:
Yeah, without interface, we kept receiving errors that no route to network existed.

pchaigno reviewed (commented):
I'm really not confortable with the amount of logic introduced here to solve something that should be relatively trivial. Starting a discussion below :point_down: to see if there's any other way.

smagnani96 committed:
cli:connectivity: json decode for pod-to-pod-encryption-v2

This commit moves our current parsing logic in pod-to-pod-encryption-v2
from strings to json decode in go structures. This makes our impl a bit
more clear.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

pchaigno commented on cilium-cli/connectivity/tests/encryption_v2.go:
```diff
@@ -114,27 +114,35 @@ func (s *podToPodEncryptionV2) Name() string {
 		srcHostNS.Pod.Name,
 		"",
-		[]string{"ip", "route", "get", dstIP})
+		[]string{"ip", "-j", "route", "get", dstIP, "from", srcIP, "iif", "cilium_host"})
```
Nit: I believe the `-j` here is in the first commit when it should be in the second.

smagnani96 replied:
You're right, nice catch, thanks :100:

pchaigno reviewed (approved):
Way easier to review now :hand_over_mouth:
